crates/subspace-networking/src/shared.rs

```diff
@@ -40,6 +43,52 @@ pub(crate) enum Command {
         message: Vec<u8>,
         result_sender: oneshot::Sender<Result<(), PublishError>>,
     },
+    GetClosestPeers {
+        key: ExactKademliaKey,
+        result_sender: oneshot::Sender<Option<Vec<PeerId>>>,
+    },
+    PiecesByRangeRequest {
+        peer_id: PeerId,
+        request: PiecesByRangeRequest,
+        result_sender: oneshot::Sender<Result<Vec<u8>, RequestFailure>>,
+    },
+}
+
+/// Kademlia Key provider without hashing. It holds the 32-bytes hash and
+/// returns the Kademlia key without double hashing. It uses the unsafe rust
+/// code to transmute types.
+#[derive(Debug, Clone)]
+pub(crate) struct ExactKademliaKey {
+    hash: [u8; 32],
+}
+
+impl ExactKademliaKey {
+    pub fn new(hash: [u8; 32]) -> Self {
+        Self { hash }
+    }
+}
+
+impl Borrow<[u8]> for ExactKademliaKey {
+    fn borrow(&self) -> &[u8] {
+        &self.hash
+    }
+}
+
+// !!! unsafe rust code
+impl From<ExactKademliaKey> for Key<ExactKademliaKey> {
+    fn from(key: ExactKademliaKey) -> Key<ExactKademliaKey> {
+        let mut data = [0u8; 64];
+
+        data[..32].copy_from_slice(key.borrow());
+
+        unsafe { std::mem::transmute::<[u8; 64], Key<ExactKademliaKey>>(data) }
```

Comment:
The suggested code will hash the parameter that we want to avoid in this snippet. The point of using the unsafe rust is to construct the `Key` struct without this hashing. As to incorrect code, I read this code as 'create an array, rebrand it as a struct with the same memory size, return the struct', it shoudn't be different than just creating the struct and returning it as parameter from the function (via stack by the way). Please, let me know if I'm missing something because it works like expected.

Reply:
Yeah, makes sense, but current code is still not correct. I think if it's okay with @nazar-pc, We could add a new constructor to our [libp2p fork](github.com/subspace/rust-libp2p)

Reply:
I don't like the unsafe code either. I discussed that with Nazar indeed and he suggested to start an issue for the libp2p: https://github.com/libp2p/rust-libp2p/issues/2690

The second option was exactly what you suggested - to create a libp2p fork similar to addition for the new public key type.

Reply:
I disabled (with a TODO comment) the code and add a simple key creation with hashing until we patch libp2p.